microsetta_public_api/api/plotting.py

```diff
@@ -155,3 +166,108 @@ def plot_beta_alt(beta_metric, named_sample_set, sample_id=None):
 
 def plot_beta(beta_metric, named_sample_set, sample_id=None):
     raise NotImplementedError()
+
+
+def _get_pcoa_repo(dataset):
+    pcoas = stepwise_resource_getter(
+        get_resources(),
+        dataset,
+        schema.pcoa_kw,
+        'pcoa',
+    )
+    pcoa_repo = PCoARepo(pcoas.data)
+    return pcoa_repo
+
+
+def _plot_ids(ax, x, y, size, marker='.', color=None, **kwargs):
+    """Plot points, return the used color"""
+    # plot is faster than scatter, go figure...
+    return ax.plot(x, y, color=color, markersize=size, marker=marker,
+                   linestyle='None', **kwargs)[0].get_color()
+
+
+def _make_mpl_fig(series, x, y, target):
+    """given metadata, coordinates and a target, make a figure"""
+    # get all the bits organized
+    df = pd.DataFrame([], index=series.index)
+    df['col'] = series
+    df['x'] = x
+    df['y'] = y
+
+    # ax1 -> plot
+    # ax2 -> legend
+    fig, (ax1, ax2) = plt.subplots(1, 2, gridspec_kw={'width_ratios': [3, 1]},
+                                   figsize=(5, 3))
+
+    # clean up the plots
+    ax1.set_xticks([])
+    ax1.set_yticks([])
+    ax2.axis('off')
+
+    # determine the point size based on the total number of samples to plot
+    n = len(series)
+    if n < 5000:
+        background_size = 5
+    elif n < 30000:
+        background_size = 1
+    else:
+        background_size = 0.1
+
+    # plot each group, keep the name and color for the legend
+    names = []
+    colors = []
+    for name, grp in df.groupby('col'):
+        colors.append(_plot_ids(ax1, grp['x'], grp['y'], background_size))
+        names.append(name)
+
+    # plot and emphasize our target
+    target = df.loc[target]
+    colors.append(_plot_ids(ax1, target['x'], target['y'], 30, marker='*',
+                            markeredgecolor='black', markeredgewidth=1.5))
+    names.append('You')
```

Comment:
Does it make sense to cache the rest of the plot and just add the enlarged marker bit when the request is made? I am not sure how much time is used for constructing the object in comparison to rendering it.

Reply:
Possibly, but it may be more expensive (see https://stackoverflow.com/questions/45810557/pyplot-copy-an-axes-content-and-show-it-in-a-new-figure). A few hundred ms for a figure that is 10x what we need to show right now is sufficient to get figures out, and we can optimize further later if needed?

Reply:
Makes sense.